src/sagemaker/workflow/_repack_model.py

```diff
@@ -33,6 +33,97 @@
 # repacking is some short-lived hackery, right??
 from distutils.dir_util import copy_tree
 
+from os.path import abspath, realpath, dirname, normpath, join as joinpath
+from sys import stderr
+
+
+def _get_resolved_path(path):
+    """Return the normalized absolute path of a given path.
+
+    abspath - returns the absolute path without resolving symlinks
+    realpath - resolves the symlinks and gets the actual path
+    normpath - normalizes paths (e.g. remove redudant separators)
+    and handles platform-specific differences
+    """
+    return normpath(realpath(abspath(path)))
+
+
+def _is_bad_path(path, base):
+    """Checks if the joined path (base directory + file path) is rooted under the base directory
+
+    Ensuring that the file does not attempt to access paths
+    outside the expected directory structure.
+
+    Args:
+        path (str): The file path.
+        base (str): The base directory.
+
+    Returns:
+        bool: True if the path is not rooted under the base directory, False otherwise.
+    """
+    # joinpath will ignore base if path is absolute
+    return not _get_resolved_path(joinpath(base, path)).startswith(base)
+
+
+def _is_bad_link(info, base):
+    """Checks if the link is rooted under the base directory.
+
+    Ensuring that the link does not attempt to access paths outside the expected directory structure
+
+    Args:
+        info (tarfile.TarInfo): The tar file info.
+        base (str): The base directory.
+
+    Returns:
+        bool: True if the link is not rooted under the base directory, False otherwise.
+    """
+    # Links are interpreted relative to the directory containing the link
+    tip = _get_resolved_path(joinpath(base, dirname(info.name)))
+    return _is_bad_path(info.linkname, base=tip)
+
+
+def safe_members(members):
+    """A generator that yields members that are safe to extract.
+
+    It checks for bad paths and bad links.
+
+    Args:
+        members (list): A list of members to check.
+
+    Yields:
+        tarfile.TarInfo: The tar file info.
+    """
+    base = _get_resolved_path(".")
+
+    for file_info in members:
+        if _is_bad_path(file_info.name, base):
+            print(stderr, file_info.name, "is blocked (illegal path)")
+        elif file_info.issym() and _is_bad_link(file_info, base):
+            print(stderr, file_info.name, "is blocked: Symlink to", file_info.linkname)
+        elif file_info.islnk() and _is_bad_link(file_info, base):
+            print(stderr, file_info.name, "is blocked: Hard link to", file_info.linkname)
+        else:
+            yield file_info
+
```

Comment:
How is the coverage tested here? tests/unit/sagemaker/workflow/test_repack_model_script.py
I see the tests are skipped with reason, reason="""This test operates on the root file system
                            and will likely fail due to permission errors.
                            Temporarily remove this skip decorator and run
                            the test after making changes to _repack_model.py"""
Can we check with @qidewenwhen on how to make sure we have the right coverage?
This can be a follow up CR by pipelines team if there is low context.

Reply:
In addition, if possible, we can ignore the `_repack_model.py` when calculating the coverage.

And as the `_repack_model.py` duplicate the logics of `custom_extractall_tarfile`, if it's properly tested in sagemaker util unit tests, we don't need to repeat the unit test here

Reply:
Yes right we can ignore `_repack_model.py` when calculating the coverage since it has the same `custom_extractall_tarfile()` func used locally in the file.